go/private/actions/compilepkg.bzl

```diff
@@ -194,3 +202,89 @@ def emit_compilepkg(
         toolchain = GO_TOOLCHAIN_LABEL,
         execution_requirements = execution_requirements,
     )
+
+    if nogo:
+        _run_nogo(
+            go,
+            sources = sources,
+            importpath = importpath,
+            importmap = importmap,
+            archives = archives + ([cover_archive] if cover_archive else []),
+            recompile_internal_deps = recompile_internal_deps,
+            cover_mode = cover_mode,
+            cgo_go_srcs = cgo_go_srcs_for_nogo,
+            out_facts = out_facts,
+            out_log = out_nogo_log,
+            out_validation = out_nogo_validation,
+            nogo = nogo,
+        )
+
+def _run_nogo(
+        go,
+        *,
+        sources,
+        importpath,
+        importmap,
+        archives,
+        recompile_internal_deps,
+        cover_mode,
+        cgo_go_srcs,
+        out_facts,
+        out_log,
+        out_validation,
+        nogo):
+    """Runs nogo on Go source files, including those generated by cgo."""
+    inputs = (sources + [nogo, go.package_list] +
+              [archive.data.facts_file for archive in archives if archive.data.facts_file] +
+              [archive.data.export_file for archive in archives] +
+              go.sdk.tools + go.sdk.headers + go.stdlib.libs)
+    outputs = [out_facts, out_log]
+
+    args = go.builder_args(go, "nogo", use_path_mapping = True)
+    args.add_all(sources, before_each = "-src")
+    if cgo_go_srcs:
+        inputs.append(cgo_go_srcs)
+        args.add_all([cgo_go_srcs], before_each = "-src")
+    if cover_mode:
+        args.add("-cover_mode", cover_mode)
+    if recompile_internal_deps:
+        args.add_all(recompile_internal_deps, before_each = "-recompile_internal_deps")
+    args.add_all(archives, before_each = "-arc", map_each = _archive)
+    if importpath:
+        args.add("-importpath", importpath)
+    else:
+        args.add("-importpath", go.label.name)
+    if importmap:
+        args.add("-p", importmap)
+    args.add("-package_list", go.package_list)
+
+    args.add_all(archives, before_each = "-facts", map_each = _facts)
+    args.add("-out_facts", out_facts)
+    args.add("-out_log", out_log)
+    args.add("-nogo", nogo)
+
+    go.actions.run(
+        inputs = inputs,
+        outputs = outputs,
+        mnemonic = "GoNogo",
+        executable = go.toolchain._builder,
+        arguments = [args],
+        env = go.env_for_path_mapping,
+        toolchain = GO_TOOLCHAIN_LABEL,
+        execution_requirements = SUPPORTS_PATH_MAPPING_REQUIREMENT,
+        progress_message = "Running nogo on %{label}",
+    )
+
+    validation_args = go.actions.args()
+    validation_args.add("nogovalidation")
+    validation_args.add(out_validation)
+    validation_args.add(out_log)
+    go.actions.run(
+        inputs = [out_log],
+        outputs = [out_validation],
+        mnemonic = "GoNogoValidation",
```

Comment:
nit: we can call this `ValidateNogo` to be consistent with `CompilePkg`.

Reply:
What would you call the action that actually runs `Nogo`? Just `RunNogo`?

Reply:
yeah, `RunNogo` sounds good